### ISSUE TITLE
Players Moving While Downed and Starting Pistol While Downed

### ISSUE THREAD

**Issue by func-octo** [Zet0rz/nzombies#312]
In the COD Zombies games (most anyway) you could move at around 10% speed while downed in order to move to a more convenient spot to be revived in. Also I suggest giving the player a hidden form of Mule Kick while downed, then give them the starter pistol without letting them switch weapons. After being revived, remove the starter pistol and the hidden form of Mule Kick. If they still have the starting pistol and are downed, only let them use that weapon.

**Comment by minin43** [comment]
While I think moving-while-down is a little pointless and I don't care for it, I am completely for the second half of your suggestion. The closer to the original game, the better.

CoD zombies also allowed for different pistols to be used (if you had another pistol in your inventory) we might want a table of "available pistols for on-down state" but the actual starter pistol is the default.

**Comment by func-octo** [comment]
> CoD zombies also allowed for different pistols to be used (if you had another pistol in your inventory) we might want a table of "available pistols for on-down state" but the actual starter pistol is the default.

I agree. Any weapon marked as a pistol in their respective addons will be used if had, and if none had, do

> giving the player a hidden form of Mule Kick while downed, then give them the starter pistol without letting them switch weapons. After being revived, remove the starter pistol and the hidden form of Mule Kick.

If I learn enough lua, I might make an addon for this addon that has a TF2 styled unlockabke weapons kinda thing. It would allow you to use any pistols in your inventory as the on-down pistol and maybe some other things.

**Comment by minin43** [comment]
Allowing it to be customizable would be best, because not every addon will properly mark it's pistol's hold-type, and if a server owner wants to enable certain weapons as valid on-down weapons, he/she can.

**Comment by func-octo** [comment]
Just realized the Admin Pistol (a.k.a. Called "Sexyness") is a pistol xD That thing kills a panzer in 1 second.

**Comment by minin43** [comment]
Why would you ever add that gun to a config...?

**Comment by func-octo** [comment]
That is an excellent question... Honestly idk why anyone would. Just to be a special snowflake, i guess :ok_hand:

**Comment by Zet0rz** [comment]
Moving while downed is awkward because gmod doesn't have proper animations for it. The best I can imagine would be the swimming backwards animation holding a pistol. I want to experiment with it, but it might not make it past the experimentation stage.

Using pistols while downed was always something I wanted to do. It already switches to a weapon you have, but there is currently nothing preventing from switching back. I should be able to easily implement a block in the CanSwitchWeapon hook. As to getting a new pistol if you don't have any, giving to the player as a SpecialWeapon should make it possible to not overwrite weapon slots and also not be able to switch away. Just gotta make the system then work without registering the weapon first (making it impossible to get normally)

**Comment by func-octo** [comment]
Variables:
- playerHasPistol() (bool, checks if player has pistol, obviously)
- playerDowned (possibly exists already)
- playerIsRevived (bool, checks to see if player has been revived recently)

Functions
- lockPistol()

function lockPistol()
----if (playerHasPistol) then
--------Switch to pistol
--------lock weapon selection
--------if (playerIsRevived) then
------------unlock weapon selection
--------end
----end
----if (!playerHasPistol) then
--------give special weapon (starter pistol as specialwep)
--------lock wepon selection
--------if (playerIsRevived) then
------------remove special weapon (starter pistol as specialwep)
------------unlock weapon selection
--------end
----end
end

if (playerDowned) then
----lockPistol() --locks player's weapon to pistol
end

**Comment by minin43** [comment]
You should check to see what hooks already exist before suggesting new ones ;)

**Comment by Zet0rz** [comment]
There's also a fundamental issue in that code. 

> if (playerIsRevived) then

When the function runs, it runs _instantly_ and _only once_. That means if you call lockPistol() when the player gets downed, playerIsRevived (or our function, player:GetNotDowned()) is false, and the code in that if-statement is never run. It doesn't "update" since the function only runs one single instant time. That's where _hooks_ come in. 

We also already have systems in place for this. You cannot switch away or to Special Weapons (and they don't replace a weapon slot). The Revive system has functions for getting whether a player is downed. We don't need to set any hasPistol variable since we can just check on downed.

**Comment by Zet0rz** [comment]
Oh look, progress is actually happening in this
https://gfycat.com/TanRipeEagle

**Comment by func-octo** [comment]
Lol, he kicks a little high to move forward, but I'm not gonna complain because it's the default backwards swimming animation :)

**Comment by func-octo** [comment]
To keep him from kicking high (to move around) you could cut some frames out and blend the others together so he only raises his leg half as high (maybe plausible?).
